### ISSUE TITLE
"No credentials" error in API.get() method

### ISSUE THREAD

**Issue by ildar-icoosoft** [aws/aws-amplify#324]
Do you want to request a feature or report a bug? Yes

What is the current behavior?

If I call API.get() method then I get an error: "No credentials". I need to do unauthenticated request without credentials.

If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than AWS Amplify.

What is the expected behavior? Do request without credentials

Which versions of Amplify, and which browser / OS are affected by this issue? Did this work in previous versions? I use aws-amplify 0.2.7

**Comment by powerful23** [comment]
@ildar-icoosoft you are getting this error because you don't have the guest credentials provided by Cognito identity pool. Can you provide more details about your configuration including Cognito and API Gateway?

**Comment by ildar-icoosoft** [comment]
Hello @powerful23. I have a public API gateway methods, so I don't need to get credentials. I just want to send request without credentials. I did not have `aws_mandatory_sign_in` because I did not know that I could use this property. Also there is no `aws_mandatory_sign_in` in automatically generated aws-exports.js. Please give me a link to documentation that describes all available properties in aws-exports.js.

**Comment by ildar-icoosoft** [comment]
aws_mandatory_sign_in does not help (tried 'enable', true, false, 'OFF')

**Comment by powerful23** [comment]
@ildar-icoosoft so you are only using the API? Are you using the Auth with AWS Cognito?

**Comment by ildar-icoosoft** [comment]
I use Auth. But before authentication users send some requests to API

**Comment by powerful23** [comment]
Then it should return a guest credential for you. Can you turn the debug mode by putting ```window.LOG_LEVEL = 'DEBUG'``` in your code to see more details?

**Comment by ildar-icoosoft** [comment]
I think I understand what went wrong. There are 4 (!) HTTP requests to AWS API AWSCognitoIdentityService.GetId method. In my Identity Pool I've disabled access to unauthenticated identities. Is it neccessary to get guest credentials? Can aws-amplify send requests without credentials?

aws-amplify tries to get guest credentials. With window.LOG_LEVEL = 'DEBUG' there are messages in console:

```
ConsoleLogger.js:84 [DEBUG] 57:10.488 AuthClass - configure Auth
ConsoleLogger.js:84 [INFO] 57:10.648 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [DEBUG] 57:10.651 I18n - configure I18n
ConsoleLogger.js:84 [DEBUG] 57:10.653 I18n - create I18n instance
ConsoleLogger.js:84 [DEBUG] 57:10.656 I18n - ru
ConsoleLogger.js:84 [DEBUG] 57:10.658 AnalyticsClass - configure Analytics
ConsoleLogger.js:84 [INFO] 57:10.661 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [DEBUG] 57:10.664 API - configure API
ConsoleLogger.js:84 [DEBUG] 57:10.666 API - create API instance
ConsoleLogger.js:100 {[DEBUG] 57:10.668 RestClient - API Options: {…}}
ConsoleLogger.js:84 [DEBUG] 57:10.675 Storage - configure called
ConsoleLogger.js:84 [DEBUG] 57:10.677 StorageClass - configure Storage
ConsoleLogger.js:84 [DEBUG] 57:10.678 StorageClass - Do not have bucket yet
ConsoleLogger.js:84 [DEBUG] 57:10.681 StorageClass - configure Storage
ConsoleLogger.js:84 [DEBUG] 57:10.683 StorageClass - Do not have bucket yet
core.js:3683 Angular is running in the development mode. Call enableProdMode() to enable the production mode.
ConsoleLogger.js:84 [INFO] 57:10.784 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [INFO] 57:10.789 Cache - Get item: key is federatedInfo with options undefined
[INFO] 57:13.675 Cache - Get item: key is federatedInfo with options undefined
[INFO] 57:16.92 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [DEBUG] 57:16.189 AuthClass - changed from previous
ConsoleLogger.js:100 {[DEBUG] 57:16.191 AuthClass - refresh credentials error: Error: Unauthenticated access is not supported for this identity pool.
ConsoleLogger.js:84 [DEBUG] 57:16.199 AuthClass - changed from previous
ConsoleLogger.js:100 {[DEBUG] 57:16.202 AuthClass - refresh credentials error: Error: Unauthenticated access is not supported for this identity pool.
ConsoleLogger.js:84 [DEBUG] 57:16.285 AuthClass - changed from previous
ConsoleLogger.js:100 {[DEBUG] 57:16.289 AuthClass - refresh credentials error: Error: Unauthenticated access is not supported for this identity pool
[DEBUG] 57:16.318 AuthClass - changed from previous
ConsoleLogger.js:100 {[DEBUG] 57:16.322 AuthClass - refresh credentials error: Error: Unauthenticated access is not supported for this identity pool
```

**Comment by ildar-icoosoft** [comment]
with aws_mandatory_sign_in === 'enable' there is no requests to AWSCognitoIdentityService.GetId method. But I still get error: "No credentials" when I call API.get()

[DEBUG] 27:00.796 AuthClass - configure Auth
ConsoleLogger.js:84 [INFO] 27:00.920 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [DEBUG] 27:00.926 I18n - configure I18n
ConsoleLogger.js:84 [DEBUG] 27:00.929 I18n - create I18n instance
ConsoleLogger.js:84 [DEBUG] 27:00.932 I18n - ru
ConsoleLogger.js:84 [DEBUG] 27:00.935 AnalyticsClass - configure Analytics
ConsoleLogger.js:84 [INFO] 27:00.941 Cache - Get item: key is federatedInfo with options undefined
ConsoleLogger.js:84 [DEBUG] 27:00.951 API - configure API
ConsoleLogger.js:84 [DEBUG] 27:00.956 API - create API instance
ConsoleLogger.js:100 {[DEBUG] 27:00.959 RestClient - API Options: {…}}
ConsoleLogger.js:84 [DEBUG] 27:00.967 Storage - configure called
ConsoleLogger.js:84 [DEBUG] 27:00.972 StorageClass - configure Storage
ConsoleLogger.js:84 [DEBUG] 27:00.977 StorageClass - Do not have bucket yet
ConsoleLogger.js:84 [DEBUG] 27:00.981 StorageClass - configure Storage
ConsoleLogger.js:84 [DEBUG] 27:00.984 StorageClass - Do not have bucket yet
[INFO] 27:05.233 Cache - Get item: key is federatedInfo with options undefined
[INFO] 27:05.233 Cache - Get item: key is federatedInfo with options undefined

**Comment by powerful23** [comment]
Yes, If you set that to 'enable' and you are not logged in, you won't get any credentials. I think you are getting the correct guest credentials from Cognito. The library tries to refresh that guest credentials but throws error.

**Comment by ildar-icoosoft** [comment]
Is it a library bug? The library should not try to refresh credentials. And the library should not get the guest credentials.

**Comment by mlabieniec** [comment]
@ildar-icoosoft Amplify will try to retrieve Cognito guest credentials for you initially for Analytics, and also so that you can do things with it, i.e. signin/up, s3, work with AWS services. This was the initial use case for the library. 

We do have in our backlog the ability to make unauthenticated requests to API Gateway as well as the ability to turn off automatic credential refresh.

**Comment by ildar-icoosoft** [comment]
@mlabieniec Thank you! Please don't forget about extra API calls. aws-amplify gets credentials 4 times instead of 1 time. Should I create an issue?

**Comment by mlabieniec** [comment]
The amount of times it gets credentials really depends on what's enabled. I'm not seeing this happen in my app, can you share what your configuration looks like? The way it currently works is IF the credentials need to be refreshed, no matter which service or API call is made, it will refresh them then do the call, as well as only if they are expired:
https://github.com/aws/aws-amplify/blob/master/packages/aws-amplify/src/Auth/Auth.ts#L855

**Comment by ildar-icoosoft** [comment]
Credentials are not expired in my case.
Here is my config:
```
Auth: {
  userPoolId: userPoolId,
  userPoolWebClientId: appClientId,
  region: region,
  identityPoolId: identityPoolId,
  mandatorySignIn: false
},
API: {
  endpoints: [{
	name: "ApiGateway",
	endpoint: "https://dbwo3ofdqh.execute-api.us-east-1.amazonaws.com/dev",
	region: region
  }]
},
Analytics: {
  // OPTIONAL -  Amazon Pinpoint App ID
  appId: pinpointId,
  // OPTIONAL -  Amazon service region
  region: region
}
```

**Comment by swaminator** [comment]
Hi @ildar-icoosoft. I'm a PM at AWS Mobile. I'd love to chat with you about your API use cases. If you could share your email with me, or DM me on twitter (@TheSwaminator), I'd be happy to set something up.

**Comment by ildar-icoosoft** [comment]
Hi @swaminator My email is ildar@icoosoft.com

**Comment by oderza** [comment]
Is the ability to make unauthenticated requests to API Gateway available yet or that still in the backlog?

**Comment by Cris87** [comment]
+1 on knowing if we can now make unauthenticated requests, I also have some methods that I have to call before user log in and am currently getting the following error:
`API - ensure credentials error: Error: Unauthenticated access is not supported for this identity pool.`
I'm assuming this is because amplify API.get() is trying to getCredentials to add the headers but since the user is not logged in the method fails. 
Is there a way to indicate when to add the Auth headers and when not to?
If I make the call using normal request I do not get any errors, I wanted to use API but am getting a lot of errors.

**Comment by oderza** [comment]
Thanks for the tip. I did have 'Enable access to unauthenticated identities' checked. What worked for me was to clear the application local storage from the browser and now everything works fine.

What might have gone wrong?

I signed up in my app several times with the same email address and then would delete the user from the Cognito User Pool. I was doing this while testing signup flow with my real email address.

However I believe the real culprit was that I tried using the non-Amplify SDK early on and I believe there may have been some remnants of that in local storage. I saved a copy of the local storage before it was cleared so that I can investigate further if necessary.
